src/django_components/tag_formatter.py

```diff
@@ -0,0 +1,256 @@
+import abc
+import re
+from typing import List, NamedTuple
+
+from django.template import TemplateSyntaxError
+from django.utils.module_loading import import_string
+
+from django_components.app_settings import app_settings
+from django_components.expression import resolve_string
+from django_components.template_parser import VAR_CHARS
+from django_components.utils import is_str_wrapped_in_quotes
+
+TAG_RE = re.compile(r"^[{chars}]+$".format(chars=VAR_CHARS))
+
+
+class TagResult(NamedTuple):
+    """The return value from `TagFormatter.parse()`"""
+
+    component_name: str
+    """Component name extracted from the template tag"""
+    tokens: List[str]
+    """Remaining tokens (words) that were passed to the tag, with component name removed"""
+
+
+class TagFormatterABC(abc.ABC):
+    def start_block_tag(self, name: str) -> str:
+        """Formats the start tag of a block component."""
+        return self.start_tag(name)
+
+    def start_inline_tag(self, name: str) -> str:
+        """Formats the start tag of an inline component."""
+        return self.start_tag(name)
+
+    @abc.abstractmethod
+    def start_tag(self, name: str) -> str:
+        """Formats the start tag of a component."""
+        ...
+
+    @abc.abstractmethod
+    def end_tag(self, name: str) -> str:
+        """Formats the end tag of a block component."""
+        ...
+
+    @abc.abstractmethod
+    def parse(self, tokens: List[str]) -> TagResult:
+        """
+        Given the tokens (words) of a component start tag, this function extracts
+        the component name from the tokens list, and returns `TagResult`, which
+        is a tuple of `(component_name, remaining_tokens)`.
+
+        Example:
+
+        Given a component declarations:
+
+        `{% component "my_comp" key=val key2=val2 %}`
+
+        This function receives a list of tokens
+
+        `['component', '"my_comp"', 'key=val', 'key2=val2']`
+
+        `component` is the tag name, which we drop. `"my_comp"` is the component name,
+        but we must remove the extra quotes. And we pass remaining tokens unmodified,
+        as that's the input to the component.
+
+        So in the end, we return a tuple:
+
+        `('my_comp', ['key=val', 'key2=val2'])`
+        """
+        ...
+
+
+class InternalTagFormatter:
```

Comment:
TagFormatter classes in this file:
- `TagResult` - The named tuple that users are expected to return from `TagFormatter.parse()`
- `TagFormatterABC` - The TagFormatter interface.
- `InternalTagFormatter` - When calling the user-provided methods, we validate that the generated tag names are valid identifiers. `InternalTagFormatter` encapsulates this validation.
- `SimpleTagFormatter` - Generic implementation of the `tag/endtag` style tags
- `ComponentFormatter` - The default behaviour with `component/endcomponent` tags
- `ShorthandComponentFormatter` - The `comp_name/endcomp_name` style tags.
	- Note the difference: While `SimpleTagFormatter` and `ComponentFormatter` define a single tag for all components, `ShorthandComponentFormatter` defines a different tag for each component

Reply:
In the docstring for the code below it looks like SimpleTagFormatter and ShorthandComponentFormatter does the exact same thing.

Also, the README omitts SimpleTagFormatter as the main defined tags.

Can you explain?

Reply:
Not the same thing, `SimpleTagFormatter` takes uses the TAG name, while `ShorthandComponentFormatter` uses the COMPONENT name.

I defined `SimpleTagFormatter` initially for the version that supported TagFormatters also for slots and fills. But kept it afterwards, because `ComponentFormatter` technically uses the same logic. But `SimpleTagFormatter` was not meant to be public.

I removed it in https://github.com/EmilStenstrom/django-components/pull/572/commits/73a748194c829fdf7c0865d009adff035ff2ff0b to avoid confusion